collectors/apiccoopdb.py

```diff
@@ -1,62 +1,40 @@
+from StandardCollector import StandardCollector, CustomMetric
 import logging
-import BaseCollector
-from prometheus_client.core import GaugeMetricFamily, Summary
+from prometheus_client.core import GaugeMetricFamily
 from typing import List, Dict
 
 LOG = logging.getLogger('apic_exporter.exporter')
-REQUEST_TIME = Summary('apic_coop_processing_seconds',
-                       'Time spent processing request')
 
 
-class ApicCoopDbCollector(BaseCollector.BaseCollector):
+class ApicCoopDbCollector(StandardCollector):
 
     def __init__(self, config: Dict):
-        super().__init__(config)
-        self.__metric_counter = 0
+        metrics = [CustomMetric(
+          name='network_apic_coop_records_total',
+          query='/api/node/class/fabricNode.json?query-target-filter=eq(fabricNode.role,"spine")',
+          process_data=self.collect_apic_coop_db_size
+        )]
+        super().__init__('apic_coop', config, metrics)
 
     def describe(self):
         yield GaugeMetricFamily('network_apic_coop_records_total',
                                 'APIC COOP DB entries')
 
-    @REQUEST_TIME.time()
-    def collect(self):
-        LOG.debug('Collecting APIC coop db metrics ...')
-
-        self.__metric_counter = 0
-
-        yield self.collect_apic_coop_db_size()
-
-        LOG.info('Collected %s APIC COOP DB metrics', self.__metric_counter)
-
-    def collect_apic_coop_db_size(self) -> List[GaugeMetricFamily]:
+    def collect_apic_coop_db_size(self, host: str, data: Dict) -> List[GaugeMetricFamily]:
```

Comment:
Why doing a different naming here? This could be a simple `collect()`, since that's what it's basically doing. Having this implementation means, that you always refer to function name `xyz` in `process_data`. At least I would go with a predefined name like: `provide_metric_data` or `generate_metrics`. Then directly use it in the StandardCollector like this.

Reply:
I agree for this simple example with just one metric to collect this would be ideal. But there are also other collectors, which collect more than one metric. For example [ApicHealthCollector](https://github.com/sapcc/apic-exporter/blob/master/collectors/apichealth.py) would require multiple functions to collect the different metrics. Since all of them also rely on different queries to retrieve the necessary data.

Using the structure to pass multiple pairs of metric and query is a compromise to group similar metrics inside one collector.